peps/pep-0728.rst

```diff
@@ -0,0 +1,640 @@
+
+This adds more corner cases to determine whether a type should be treated as a
+type or a value. And it will require more work to support using special forms to
+type the extra items.
+
+While this saves us from reserving an attribute for special use, it will require
+extra work to implement inheritance, and it is less natural to integrate with
+generic TypedDicts.
+
+Support Extra Items with Intersection
+-------------------------------------
+
+Supporting intersections in Python's type system requires a lot of careful
+considerations, and it can take a long time for the community to reach a
+consensus on a reasonable design.
+
+Ideally, extra items in TypedDict should not be blocked by work on
+intersections, nor does it necessarily need to be supported through
+intersections.
+
+Moreover, the intersection between ``Mapping[...]`` and ``TypedDict`` is not
+equivalent to a TypedDict type with the proposed ``__extra__`` special item, as
+the value type of all known items in ``TypedDict`` needs to satisfy the
+is-subtype-of relation with the value type of ``Mapping[...]``.
+
+Requiring Type Compatibility of the Known Items with ``__extra__``
+------------------------------------------------------------------
+
+``__extra__`` restricts the value type for keys that are *unknown* to the
+TypedDict type. So the value type of any *known* item is not necessarily
+consistent with ``__extra__``'s type, and ``__extra__``'s type is not
+necessarily consistent with the value types of all known items.
+
+This differs from TypeScript's `Index Signatures
+<https://www.typescriptlang.org/docs/handbook/2/objects.html#index-signatures>`__
+syntax, which requires all properties' types to match the string index's type.
+For example:
+
+.. code-block:: typescript
+
+    interface MovieWithExtraNumber {
+        name: string // Property 'name' of type 'string' is not assignable to 'string' index type 'number'.
+        [index: string]: number
+    }
+
+    interface MovieWithExtraNumberOrString {
+        name: string // OK
+        [index: string]: number | string
+    }
+
+This is a known limitation is discussed in `TypeScript's issue tracker
+<https://github.com/microsoft/TypeScript/issues/17867>`__,
+where it is suggested that there should be a way to exclude the defined keys
+from the index signature, so that it is possible to define a type like
+``MovieWithExtraNumber``.
+
+Reference Implementation
+========================
+
+pyanalyze has `experimental support
+<https://github.com/quora/pyanalyze/blob/9bfc2c58467c87774a9950838402d2657b1486a0/pyanalyze/extensions.py#L590>`__
+for a similar feature.
+
+Reference implementation for this specific proposal, however, is not currently
+available.
+
+Acknowledgments
+===============
+
+Thanks to Jelle Zijlstra for sponsoring this PEP and providing review feedback,
+Eric Traut who `proposed the original design
+<https://mail.python.org/archives/list/typing-sig@python.org/message/3Z72OQWVTOVS6UYUUCCII2UZN56PV5II/>`
```

Comment:
This isn't rendering as a link properly, I think it needs a  couple of underscores at the end?